pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -1100,6 +1170,638 @@ class StateVectorCudaManaged
         return -static_cast<PrecisionT>(0.5);
     }
 
+    /* Controlled-gate generators */
+    /**
+     * @brief Gradient generator function associated with the controlled-RX
+     * gate.
+     *
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (false or true).
+     * @param wires Target wires to apply operation.
+     * @param adj Takes adjoint of operation if true. Defaults to false.
+     */
+    inline PrecisionT
+    applyControlledGeneratorRX(const std::vector<std::size_t> &controlled_wires,
+                               const std::vector<bool> &controlled_values,
+                               const std::vector<std::size_t> &wires,
+                               bool adj = false) {
+        const std::size_t ctrl_size = controlled_wires.size();
+        const std::size_t tgt_size = wires.size();
+
+        // Generate permutation
```

Comment:
Could we move the code for permutation generations and diagonals generations to a private method for better maintenance?

Reply:
I can move them to util

Reply:
I have refactored the permutation generation and index calculation into a helper function; I don't want to combine diagonal together because in cases where there is no permutation, I can pass an empty permutation vector to the custatevec call, so there is no need to compute/generate the permutation vector.